examples/basic-federation/server.go

```diff
@@ -0,0 +1,73 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/smallstep/certificates/ca"
+)
+
+func main() {
+	if len(os.Args) != 2 {
+		fmt.Fprintf(os.Stderr, "Usage: %s <token>\n", os.Args[0])
+		os.Exit(1)
+	}
+
+	token := os.Args[1]
+
+	// make sure to cancel the renew goroutine
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	srv, err := ca.BootstrapServer(ctx, token, &http.Server{
+		Addr: ":8443",
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			name := "nobody"
+			issuer := "none"
+			if r.TLS != nil && len(r.TLS.PeerCertificates) > 0 {
+				name = r.TLS.PeerCertificates[0].Subject.CommonName
+				issuer = r.TLS.PeerCertificates[len(r.TLS.PeerCertificates)-1].Issuer.CommonName
+			}
+
+			w.Write([]byte(fmt.Sprintf("Hello %s (cert issued by '%s') at %s", name, issuer, time.Now().UTC())))
+		}),
+	}, ca.AddFederationToRootCAs(), ListTrustedRoots())
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Println("Listening on :8443 ...")
+	if err := srv.ListenAndServeTLS("", ""); err != nil {
+		panic(err)
+	}
+}
+
+// ListTrustedRoots prints list of trusted roots for illustration purposes
+func ListTrustedRoots() ca.TLSOption {
+	fn := func(ctx *ca.TLSOptionCtx) error {
+		certs, err := ctx.Client.Federation()
+		if err != nil {
+			return err
+		}
+		roots, err := ctx.Client.Roots()
+		if err != nil {
+			return err
+		}
+
+		if len(certs.Certificates) > len(roots.Certificates) {
+			fmt.Println("Server is using federated root certificates")
```

Comment:
This is weird, the server is using both, federated and root

Reply:
You mean the wording of the log message?

Reply:
Yes, but I suppose for the example is ok.
I added the comment because in the requests before the comment (get federation and roots), the certs used are the root ones.